### ISSUE TITLE
Legacy branches, PRs and issues

### ISSUE THREAD

**Issue by roman-khimov** [nspcc-dev/neo-go#283]
After the project transfer and before we start the new development we need to clean up the repository. IMO it's better to do it in an open and transparent way thus this issue. Some changes were made yesterday, but that were things like deleting already merged branches or unused projects/milestones, supposedly harmless things. Now I'd really like to get this repository into some reasonable clean state and that would require making more invasive changes so if there are any suggestions or objections for that changes --- this is the place to discuss it.

First of all, at the moment we have six branches, namely:
 * `master`
 * `dev`
 * `v2dev`
 * `blockPool`
 * `vm`
 * `interop-test-dummy`

The most annoying problem here is that the `dev` branch (and its descendants) has completely separate root from the `master` branch (and its descendants), so that their history have nothing in common from the git point of view. Luckily, the initial commit on the `dev` branch (cda7a31e4e0b1494652661f2fd8ce07a79629b75) actually copies the state of the latest commit from the `v2dev` branch (101ecf907695952b2371db8fdc0f1035bcc1fab2), which actually has some common history with `master` so there is a possibility of consolidation via rebase of the `dev` changes (from 7eafd1ac17909e25e516f6251c9df9a698790a25 onward) to the `v2dev` branch. But that should probably be done as the last step of the cleanup because it's the most invasive and debatable change. The `v2dev` branch is only interesting as a starting point for `dev`, nothing more.

All the other branches are `dev` descendants, `blockPool` contains changes outlined in the PR #264, `vm` contains lots of changes to the virtual machine and `interop-test-dummy` has just one little patch with an attempt to add the VM test from the neo-vm repository. IMO all of these branches should just be merged into the `dev` to consolidate the changes we already have.

Same goes for open PRs like #261, #265, #266, #267, #271, #273, #275, #277, #279, all of these are VM improvements made on top of `vm` (which in turn is made on top of `dev`). I think we can also merge PR #120 into the `master` branch, although it's just a stub, but it will serve as a reminder.

This would get us to the point where we gonna have two branches: `master` and `dev`, with all the precious changes made and with all the history preserved and then we could start dealing with differences between those as they are basically two completely different beasts at the moment.

Summarizing, what I propose to do is:
- [x] merge PR #264 (as is) and drop `blockPool` branch
- [x] merge `interop-test-dummy` into `vm` (as is) and drop it
- [x] merge PRs #261, #265, #266, #267, #271, #273, #275, #277, #279 (as is)
- [x] merge PR #120 (as is)
- [x] merge `vm` into the `dev` and drop it
<del>- [ ] rebase `dev` on top of `v2dev` and drop `v2dev`</del>

Next I'm going to look into the issues list with two basic questions: are they still relevant and if so then which branch they're applicable to.

**Comment by roman-khimov** [comment]
One additional note to the dev/master problem --- technically, if we're to decide to base our work on top of one of these branches (and we'll have to choose one anyway), the other one could be deleted, so that's an argument against rebasing `dev` on top of `v2dev` (just not to waste time doing that).

**Comment by im-kulikov** [comment]
I think, we have some troubles with merging v2dev (<-dev) into master. There's no JSON-RPC realization (fix me if it’s not true) in dev branch..
Any thoughts?

**Comment by im-kulikov** [comment]
master branch has on top features for json-rpc, thus I think, we must kept they after merging

**Comment by im-kulikov** [comment]
on the other hand, it also seems to me very important to preserve the history of commits, because this is the contribution that the community made..

**Comment by roman-khimov** [comment]
I don't propose merging `dev` and `master` at the moment, it's a separate issue. I think at first we need to aggregate all the work around these two (radically) different branches into, well, two branches (and not 6 branches with 12 PRs) and then discuss the differences between them.

Also, I'd really like to hear something from @decentralisedkev as he probably knows this codebase (or even codebases given the difference between dev and master) way better than we're at the moment.

**Comment by dauTT** [comment]
Nice to see that this project is not dead and is coming back to life under new leadership. I believe that the `dev` branch is somehow better of the `master` branch in term of modularity. This is why @decentralisedkev  choose to stop develop `master` branch and gave great emphasis on working on the `dev` branch. There was also discussion in the past to make the `master` branch more modular but the effort to refactor it was too big that in the end it was decided that it was much easier to rewrite the code from scratch. Therefore community contributors started to focus more on the `dev` branch.

**Comment by decentralisedkev** [comment]
Hi,

**Branches confusion**

Thinking back, the main reason why the master branch is still around is that others have contributed to it.
`v2dev` shares history with the master branch, as my first idea was to remove all files and commit. This kept the previous contributions, however, it made the history somewhat confusing as the modular approach was a completely new codebase, and the previous approach had no working code.
After a couple of months of inactivity, I created the `dev` branch with the modular approach from an empty fit history. This will therefore not be mergeable into master. What I can propose is that you rename master to `deprecated-code`, make dev the new master and remove `v2dev`.
If the git history confusion is not a problem, then you can also rebase on top of master.

**Note about RPC**

I’m not sure if the RPC module can be copied over from `master` to `dev`. You can check by observing if doing this causes the modularity principle to break down.  

**Current state of project**:

The node should be able to sync with one other node and download the blockchain up to the most recent block. Without verifying the block.

**Envisioned Roadmap**

_Parallel sync_
Although syncing with one client was reasonably fast, before I stopped developing, we were going to push a release with parallel synchronisation. This was completed. If I can find it on my laptop, I will push to a branch on my fork and you can decide whether to merge.
_VM_
The structure for the VM has been completed. The main task at the moment is adding all of the opcodes which takes a bit of time and is quite repetitive. There are some differences with neo-python, I believe I made some issues for those.
I was looking into a way to conveniently use the test files in the neovm to test interoperability.

_Crypto_
The crypto modules from the master branch are unsafe. The plan was to re-implement the curve in golang; field elements, curve models, etc as the big.Int package is not safe for crypto purposes. 
If time is of the essence for you, then you can also opt to use the big.Int package, however noting that it is not safe as there are variable time operations is what is conventionally done in the industry.

If you have any questions or doubts about any decisions I made, please feel free to ask or modify to fit your specs as you see fit.

**Comment by decentralisedkev** [comment]
I have gone through the issues, and most can be inferred from the project structure except the wallet.

The wallet was going to be separate from the node itself. I did not finish the specs to implement the merkle patricia trie, making this approach in-efficient. Since there is no documentation on this, and the codebase shows no bias towards it at the moment. It can be implemented in any way you see fit.

**Comment by roman-khimov** [comment]
Thanks for your input. I think we should proceed with branch merges as outlined above except the last point with `dev` rebase (there is another argument against that --- 91a1e88cd894653698deddd24fccc915acd2fc3d which makes previous history almost useless). Let's schedule this part for Monday 12th (around 9am UTC), so that there is still some chance for everybody to weigh in on this topic before that.

**Comment by roman-khimov** [comment]
> The crypto modules from the master branch are unsafe.

Maybe we can move this part of discussion into #81?

**Comment by roman-khimov** [comment]
Linking #113 although there is not that much info there.

**Comment by roman-khimov** [comment]
OK, so now we have `dev`, `v2dev` and `master` branches. I tried rebasing `dev` on top of `v2dev`, but it didn't work well and will require quite a lot of conflict resolutions. Given that `v2dev` history is a mess anyway (like 91a1e88cd894653698deddd24fccc915acd2fc3d) rebasing just won't be much of an improvement. So, we'll let `dev` and `master` be disconnected for a while, and `v2dev` can just be dropped as not having any real value (all the code is in the `dev` anyway).

After that I'm thinking about merging `master` into `dev` thus forming one development branch (to be renamed into `master`). So all the `dev` code is not going to be changed, but we will still pick things from master (with all the history!) that are just completely missing in `dev` (cli, config, rpc and wallet are good examples but there are probably more). The end result would probably not work at the beginning (as the old code would need some adaptation), but we will gradually solve that issue.

Summarizing, the next actions are:
- [ ] drop `v2dev`, scheduled for tomorrow (Wednesday 14th) after 18:00 UTC
- [ ] merge `master` into `dev`, scheduled for Thursday 15th after 12:00 UTC

I think it gives enough time for anyone to step in with suggestions/objections to this plan.

**Comment by roman-khimov** [comment]
On `master` and `dev` differences again. I think `dev` is comparable to `master` in terms of functionality (of course, not counting things like #214 and #186) for the following packages (directories):
* pkg/core
* pkg/crypto
* pkg/io
* pkg/network
* pkg/smartcontract
* pkg/vm (with exceptions)

so we can just drop these on merge. But it lacks the following things (that would be nice to merge):
* cli
* config
* pkg/rpc
* pkg/vm/cli
* pkg/vm/compiler (see #290 also)
* pkg/wallet

which we should keep. It also lacks some opcodes implementation in VM, but that shouldn't be hard to fix.

Maybe @anthdm can add something to this discussion? Are there any other known missing functionality in `dev` that is present in `master`? I want to minimize regressions moving to dev.

**Comment by roman-khimov** [comment]
`v2dev` is gone now and I'm rescheduling master/dev merge to Monday 19th because of #290, it all takes time.

**Comment by roman-khimov** [comment]
So, we have two new branches: `devaster`, which is a master merged to `dev` and `mastev`, which is a `dev` merged to `master`. And after several additional discussions I think that the `mastev` approach is better, that is, we're basing our future work on top of the `master` branch and bring the code from `dev` when and where it'll be needed. The rationale is:
 * `master` is way more feature-complete, porting things from it into `dev` will take more time than the other way around
 * `dev` has overly complicated structure with managers managing managers, functions being passed through channels for no good reason and things like that (that's opinionated, of course)
 * `master` has better practical test suite, again, it's easier to port some tests from `dev`

So I'll wait for feedback till tomorrow (Wednesday 21st) 16:00 UTC and if there are no objections then I will:
 * drop `devaster`
 * rename `mastev` to `master`
 * make `master` the default branch
 * drop `dev`
 * close issues irrelevant for `master`
 * add issues to port things from `dev` where appropriate
 * close this issue

Then we'll proceed with roadmaps/release schedule and things like that.

Obviously, `_pkg.dev` is a kludge, but it'll go away when all the nice pieces from `dev` are ported into `master`.

**Comment by roman-khimov** [comment]
OK, so now we no longer have obsolete branches and PRs, all of the issues are sorted out and either confirmed or closed, so this one is solved (hooray!) and should be closed. I'm also going to drop `affectsMaster` and `affectsDev` tags that were useful during issue sorting phase but now just lost their meaning as all the open issues are relevant for our future work.
